tools/cd_scripts/e2e_test.sh

```diff
@@ -139,28 +140,192 @@ git checkout $(sed -n 2p ~/details.txt) |& tee -a ~/logs.txt
-if [ $? -ne 0 ];
-then
-    echo "Test failures detected" &>> ~/logs-hns.txt
-else
-    touch success-hns.txt
-    gsutil cp success-hns.txt gs://gcsfuse-release-packages/v$(sed -n 1p ~/details.txt)/$(sed -n 3p ~/details.txt)/
-fi
-gsutil cp ~/logs-hns.txt gs://gcsfuse-release-packages/v$(sed -n 1p ~/details.txt)/$(sed -n 3p ~/details.txt)/
+# These tests never become parallel as they are changing bucket permissions.
+TEST_DIR_NON_PARALLEL=(
+  "readonly"
+  "managed_folders"
+  "readonly_creds"
+  "list_large_dir"
+)
+
+# Create a temporary file to store the log file name.
+TEST_LOGS_FILE=$(mktemp)
+
+GO_TEST_SHORT_FLAG="-short"
+INTEGRATION_TEST_TIMEOUT=300m
+
+function run_non_parallel_tests() {
+  local exit_code=0
+  local -n test_array=$1
+  local BUCKET_NAME=$2
+  for test_dir_np in "${test_array[@]}"
+  do
+    test_path_non_parallel="./tools/integration_tests/$test_dir_np"
+    # To make it clear whether tests are running on a flat or HNS bucket, We kept the log file naming
+    # convention to include the bucket name as a suffix (e.g., package_name_bucket_name).
+    local log_file="/tmp/${test_dir_np}_${BUCKET_NAME}.log"
+    echo $log_file >> $TEST_LOGS_FILE
+    # Executing integration tests
+    GODEBUG=asyncpreemptoff=1 go test $test_path_non_parallel -p 1 $GO_TEST_SHORT_FLAG --integrationTest -v --testbucket=$BUCKET_NAME --testInstalledPackage=true -timeout $INTEGRATION_TEST_TIMEOUT > "$log_file" 2>&1
+    exit_code_non_parallel=$?
+    if [ $exit_code_non_parallel != 0 ]; then
+      exit_code=$exit_code_non_parallel
+    fi
+  done
+  return $exit_code
+}
+
+function run_parallel_tests() {
+  local exit_code=0
+  local -n test_array=$1
+  local BUCKET_NAME=$2
+  local pids=()
+  for test_dir_p in "${test_array[@]}"
+  do
+    test_path_parallel="./tools/integration_tests/$test_dir_p"
+    # To make it clear whether tests are running on a flat or HNS bucket, We kept the log file naming
+    # convention to include the bucket name as a suffix (e.g., package_name_bucket_name).
+    local log_file="/tmp/${test_dir_p}_${BUCKET_NAME}.log"
+    echo $log_file >> $TEST_LOGS_FILE
+    # Executing integration tests
+    GODEBUG=asyncpreemptoff=1 go test $test_path_parallel $GO_TEST_SHORT_FLAG -p 1 --integrationTest -v --testbucket=$BUCKET_NAME --testInstalledPackage=true -timeout $INTEGRATION_TEST_TIMEOUT > "$log_file" 2>&1 &
+    pid=$!  # Store the PID of the background process
+    pids+=("$pid")  # Optionally add the PID to an array for later
+  done
+  # Wait for processes and collect exit codes
+  for pid in "${pids[@]}"; do
+    wait $pid
+    exit_code_parallel=$?
+    if [ $exit_code_parallel != 0 ]; then
+      exit_code=$exit_code_parallel
+    fi
+  done
+  return $exit_code
+}
+
+function run_e2e_tests_for_flat_bucket() {
+  flat_bucket_name_non_parallel=$(sed -n 3p ~/details.txt)
```

Comment:
nit: Add echo for "Running tests for FLAT bucket". same for HNS

Reply:
Added.